src/pykiso/test_coordinator/assert_step_report/assert_step_report.py

```diff
@@ -96,13 +97,30 @@ def _get_variable_name(f_back: types.FrameType, assert_name: str) -> str:
     # Get current line number
     line_no = f_back.f_lineno
 
-    # Get line content, if statement not
-    # complete, read from bottom to up
+    # The inspect module returns the first line in python >=3.8 and the last line in python 3.7
+    first_line = sys.version_info >= (3, 8)
+
+    # Get line content
     line = ""
-    while assert_name not in line:
-        # line_no start from 1 but list from 0
+    if first_line:  # pragma: no cover
```

Comment:
I've made an attempt to make this function a bit more readable, it runs on 3.7 and 3.9:

```python
import linecache

def _get_variable_name(f_back: types.FrameType, assert_name: str) -> str:
    """Get the input parameter name to the assert method.

    According to unittest.TestCase, it is always placed first.
    That function read the source code as a text to find it.

    :param f_back: frame calling the assert method
    :param: assert_name: current assert method name

    return: variable name
    """
    expected_varname = ""
    # Get source code lines
    file = inspect.getsourcefile(f_back)

    # Get current line number
    line_no = f_back.f_lineno

    # For multiline statements, the current frame points to the last line in python < 3.8
    # and to the first line in python >= 3.8
    first_line = sys.version_info >= (3, 8)

    # Get line content
    line = linecache.getline(file, line_no)
    lines = line

    if first_line: # pragma: no cover
        # Read top down, count parentheses to find last line of assert statement
        open_parentheses, closed_parentheses = line.count("("), line.count(")")
        while open_parentheses != closed_parentheses: 
            line_no += 1
            line = linecache.getline(file, line_no)
            lines += line
            open_parentheses += line.count("(")
            closed_parentheses += line.count(")")
    else:
        # Read bottom up until the assert method is reached
        while assert_name not in line:
            line_no -= 1
            previous_line = linecache.getline(file, line_no)
            lines = previous_line + line
            line = previous_line

    # remove line breaks and spaces
    lines = lines.replace("\n", "").replace(" ", "")

    # remove current assert function name
    lines = re.split(rf".*{assert_name}\(", lines)[1]

    # Check the variable start names
    if re.findall(r"^[a-zA-Z]", lines):
        # Get variable name
        expected_varname = re.findall(r"[a-zA-Z0-9_]+", lines)[0]

    return expected_varname
    ```

Reply:
It looks a little bit better, I will copy it.